### DISCUSSION TITLE
Request: TypeScript awareness

### DISCUSSION THREAD

**Discussion by ffxsam** [community/community#6944] [Copilot Conversations] [answered]
I noticed oftentimes Copilot will make suggestions that don't make any sense, even though it could feasibly use type awareness to make certain determinations when suggesting code fragments.

**Comment by Zatef1** [comment]
you're so lucky even to have copilot and you're complaining about " having some nonsense suggestions!" I have been waiting for a while to get the opportunity to even have copilot and just not yet:/

**Reply by ffxsam** [reply]
This is the place where Copilot beta testers give feedback. Relax.

**Reply by devAgam** [reply]
Github Co-pilot is in beta, this is a beta tester feedback discussion, so chill out.

**Comment by johanrosenkilde** [comment, marked as answer]
Thanks for your feedback! Giving Copilot better type understanding is something we're very actively working on: the technology behind Copilot is very different from classical program analysis, and so it is not straightforward how to "tell" Copilot what types things have.
However, in TypeScript the following should work already: If you're working in a file which imports an interface from another file in your package, then Copilot should generally pick up on available properties for variables of this type.

**Reply by ffxsam** [reply]
Unfortunately, it does make false suggestions even though I have an enum that's clearly defined:

Copilot suggested this:

<img width="324" alt="image" src="https://user-images.githubusercontent.com/12532733/139463284-6a64571a-330b-417e-945b-2eeebcd2aec9.png">

And yet:

```ts
export enum ItemType {
  TRACK = 'TRACK',
  IMAGE = 'IMAGE',
  VIDEO = 'VIDEO',
  REEL = 'REEL',
  SNIPPET = 'SNIPPET',
}
```

**Reply by johanrosenkilde** [reply]
Yes, this is still work-in-progress - it's currently only active for *interfaces*, not types or enums.

**Reply by ffxsam** [reply]
Ah, gotcha. Thanks, Johan!

**Comment by AlexDM0** [comment]
Just wondering, it does not seem to work with interfaces either. Is this still work in progress?

See type definition and suggestion
<img width="256" alt="Screenshot 2022-06-30 at 13 47 03" src="https://user-images.githubusercontent.com/5363277/176670018-80c1a519-fa64-417c-83da-b6131a462269.png">
<img width="447" alt="Screenshot 2022-06-30 at 13 46 57" src="https://user-images.githubusercontent.com/5363277/176670022-ad5d28c1-b9dc-49b1-a17b-7f3391c1d428.png">

I've just started using copilot after the public release.

It does not matter if the interface definition is in a definitions.d.ts file or in the same file as the function it is suggestion.

**Reply by johanrosenkilde** [reply]
It should definitely work, but Copilot may still ignore it and suggest properties outside the type. In your case, I'm not even sure as a human what you'd want `getAmountOfDevicesInData` to sum over, as in, what your type represents is not clear from the names. Copilot's suggestion doesn't seem unreasonable to me, as in, perhaps you forgot to say when a datum is a device type or not. You could try adding a comment to the type, or you could just keep coding in the current file, and Copilot should pick up on the type's intention from how you're using it.

**Reply by AlexDM0** [reply]
That's fair, I just quickly changed the keywords for the example. However, the result will lead to a typescript compilation error. I figured that would weigh quite heavily in deciding what to suggest. There is no .deviceType element on the data. Thanks for the reply!
